core/rawdb/freezer.go

```diff
@@ -269,17 +269,24 @@ func (f *freezer) freeze(db ethdb.KeyValueStore) {
 			select {
 			case <-time.NewTimer(freezerRecheckInterval).C:
 				backoff = false
-			case <-f.quit:
+			case <-quitChan:
 				return
 			}
 		}
+
 		// Retrieve the freezing threshold.
 		hash := ReadHeadBlockHash(nfdb)
 		if hash == (common.Hash{}) {
 			log.Debug("Current full block hash unavailable") // new chain, empty database
 			backoff = true
 			continue
 		}
+
+		numFrozen, err := f.Ancients()
```

Comment:
Just a call for :eyes: . 

Using an assigned value like this, rather than having direct access to the 'live' `f.frozen` field, is a quite drastic logical step. We need to be sure that the value will always be updated in sync with the subsequent logic depending on it, and that we're never going to be thrashing the RPC in this loop.

Reply:
And this modifies the built-in FS ancient database mechanism. So, double double check.

Reply:
An alternative here that may actually be preferable would be to duplicate the logic into

```go
// original FS implementation, using f.frozen and unmodified function signature
func freezeLocal

// freezer client implementation, using f.Ancients() and modified function signature
func freezeRemote
```

`NewDatabaseWithFreezer` and `NewDatabaseWithFreezerRemote` are the only two callers, so the refactoring and code dependencies are simple.

Reply:
I agree with that then we can reduce the surface area of things that can go wrong and prove out that ancient remote works as expected over a longer time frame.

Reply:
https://github.com/etclabscore/core-geth/pull/168/commits/f528c61da5963f9eb26442a88f5ce001fc474097 implements [this proposal](https://github.com/etclabscore/core-geth/pull/168#discussion_r465894730).

Reply:
yeah I think this makes the most sense for now, it gives us some space to prove out the remote freezer logic. any thoughts on keeping track of any critical upstream patches to freezer within remote freezer?

Reply:
I mean, to me the code before https://github.com/etclabscore/core-geth/commit/f528c61da5963f9eb26442a88f5ce001fc474097 looked OK too (with `f.frozen` using `f.Ancients()` instead). The only reason for the refactoring to near-dupe functions is to reduce the risk in case I'm wrong. As @zcstarr pointed out however, duplication introduces an overhead in code-life audits and maintenance.

Reply:
for the record here, I'd say the code looks correct for f.Ancients() as well.